python/cudf/cudf/tests/utils.py

```diff
@@ -77,7 +77,13 @@ def assert_eq(left, right, **kwargs):
         left = cupy.asnumpy(left)
     if isinstance(right, cupy.ndarray):
         right = cupy.asnumpy(right)
-
+    if (
+        allow_nullable_pd_types
+        and isinstance(left, (pd.Series, pd.DataFrame))
+        and left.__class__ == right.__class__
+    ):
+        left = maybe_demote_dtypes(left)
+        right = maybe_demote_dtypes(right)
```

Comment:
This feels like a very error prone approach. Why do we need to demote the types to check for equality?

Reply:
This is something I was hoping to discuss - the problem is that our testing relies on using `to_pandas` to drop to a pandas object before asserting equality. If we're changing `to_pandas()` such that we get nullable dtypes instead of numpy dtypes, that means that pandas objects and cudf objects that should compare as equal no longer will. For example if I have the cuDF series `['a', null, 'c'], dtype='object'` and the pandas series `['a', None, 'c'], dtype='object'`, these will no longer compare as equal because when we call `to_pandas` on the cuDF series, we will get a `pd.StringDtype()` in the output instead of coercing to object like we previously did. 

We either need to promote the pandas result ("expect" in our tests) to nullable dtypes, or demote the cuDF `to_pandas()` output down to numpy dtypes before comparing them. Or, rework all of our tests so that the "expect" data on the pandas side uses extension dtypes from the getgo, but I figured we'd probably want to discuss what to do before I update the vast majority of our tests - and if we do go that route we might run into issues where our null handling/propagation behavior differs from `pd.NA` in subtle ways. 

Because of this it struck me as the simplest solution to just loosen our requirements for equality to just say that a pandas series with a nullable dtype is equal to another pandas series with the non-nullable equivalent data.

Reply:
To me option 1, i.e.,

> promote the pandas result ("expect" in our tests) to nullable dtypes

sounds the most attractive. If we did this, would our tests just pass?

Reply:
IIRC there were some issues with doing that (I think that's what I tried originally and switched it for some reason) but seeing as how I can't actually remember exactly what went wrong I'll try it and see what falls out.